src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -250,28 +255,79 @@ public override async ValueTask OnReferenceFoundAsync(SourceReferenceItem refere
+                    var classifiedTextRuns = GetClassifiedTextRuns(id, definitionId, documentSpan, isWrittenTo, classifiedSpans, docText);
 
-                    return new ClassifiedTextElement(
-                        classifiedSpans.Select(cspan => new ClassifiedTextRun(cspan.ClassificationType, docText.ToString(cspan.TextSpan))));
+                    return new ClassifiedTextElement(classifiedTextRuns.ToArray());
+                }
+                // Certain definitions may not have a DocumentSpan, such as namespace and metadata definitions
+                else if (id == definitionId)
+                {
+                    return definitionText;
                 }
 
                 return null;
+
+                // Nested local functions
+                static ClassifiedTextRun[] GetClassifiedTextRuns(
+                    int id, int? definitionId,
+                    DocumentSpan documentSpan,
+                    bool isWrittenTo,
+                    ImmutableArray<ClassifiedSpan> classifiedSpans,
+                    SourceText docText)
+                {
+                    using var _ = ArrayBuilder<ClassifiedTextRun>.GetInstance(out var classifiedTextRuns);
+                    foreach (var span in classifiedSpans)
+                    {
+                        // Case 1: Highlight this span of text. For example, if the user invokes FAR on 'x' in 'var x = 1',
+                        // then 'x' should be highlighted.
+                        if (span.TextSpan == documentSpan.SourceSpan)
+                        {
+                            // Case 1a: Highlight a definition
+                            if (id == definitionId)
+                            {
+                                classifiedTextRuns.Add(new ClassifiedTextRun(
+                                    span.ClassificationType, docText.ToString(span.TextSpan), ClassifiedTextRunStyle.Plain,
+                                    markerTagType: DefinitionHighlightTag.TagId));
+                            }
+                            // Case 1b: Highlight a written reference
+                            else if (isWrittenTo)
+                            {
+                                classifiedTextRuns.Add(new ClassifiedTextRun(
+                                    span.ClassificationType, docText.ToString(span.TextSpan), ClassifiedTextRunStyle.Plain,
+                                    markerTagType: WrittenReferenceHighlightTag.TagId));
+                            }
+                            // Case 1c: Highlight a read reference
+                            else
+                            {
+                                classifiedTextRuns.Add(new ClassifiedTextRun(
+                                    span.ClassificationType, docText.ToString(span.TextSpan), ClassifiedTextRunStyle.Plain,
+                                    markerTagType: ReferenceHighlightTag.TagId));
```

Comment:
these three blocks are identical (except for the tag type.  It would cleaner to just use a simple switch expr to determine the marker type, then just create/add the tag once usin gthat.